lib/guest_configuration_base.pm

```diff
@@ -0,0 +1,1408 @@
+        #guest installation screen using [guest_installation_session_command] = screen -t [guest_name] virsh console --force [guest_name]
+    'guest_installation_attached' => '',    #This indicates whether guest installation screen is already connected or attached(true or false)
+    'start_run'                   => '',    #Guest creation start time
+    'stop_run'                    => ''     #Guest creation finish time
+);
+
+our $AUTOLOAD;
+our $common_log_folder           = '/tmp/guest_installations';
+our $common_environment_prepared = 'false';
+
+#Any subroutine calls this subroutine announces its identity and it is being executed
+sub reveal_myself {
+    my $self = shift;
+
+    my $_my_identity = (caller(1))[3];
+    diag("Test execution inside $_my_identity.");
+    return $self;
+}
+
+#Create guest instance by assigning values to its parameters but do no install it
+sub create {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->initialize_guest_params;
+    $self->config_guest_params(@_);
+    $self->print_guest_params;
+    return $self;
+}
+
+#Initialize all guest parameters to avoid uninitialized parameters
+sub initialize_guest_params {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->{$_} //= '' foreach (keys %guest_params);
+    $self->{host_ipaddr} = get_required_var('SUT_IP');
+    $self->{host_name}   = script_output("hostname");
+    $self->{start_run}   = time();
+    return $self;
+}
+
+#Assign real values to guest instance parameters.Reset [guest_name] to guest name used in [guest_metadata] if they are different.
+sub config_guest_params {
+    my $self          = shift;
+    my %_guest_params = @_;
+
+    $self->reveal_myself;
+    if ((scalar(@_) % 2 eq 0) and (scalar(@_) gt 0)) {
+        record_info("Configuring guest instance by using following parameters:", Dumper(\%_guest_params));
+        map { $self->{$_} = $_guest_params{$_} } keys %_guest_params;
+    }
+    else {
+        record_info("Can not configure guest instance with empty or odd number of arguments.Mark it as failed.", Dumper(\%_guest_params));
+        $self->record_guest_installation_result('FAILED') if ($self->{guest_installation_result} eq '');
+        return $self;
+    }
+
+    if ((defined $self->{guest_metadata}) and (grep { /name=/ } split(',', $self->{guest_metadata}))) {
+        my $_guest_name_in_metadata = grep { /name=(.*)/ } split(',', $self->{guest_metadata});
+        if ($self->{guest_name} ne $_guest_name_in_metadata) {
+            diag("The guest instance has a different name $_guest_name_in_metadata in metadata. Reset guest name to it.");
+            $self->{guest_name} = $_guest_name_in_metadata;
+        }
+    }
+    if ($self->{guest_version} eq lc get_required_var('VERSION')) {
+        $self->{guest_build} = lc get_required_var('BUILD');
+    }
+    else {
+        $self->{guest_build} = 'GM';
+    }
```

Comment:
For guest_version and guest_build variables, although what you configured are the most common case which really applied, but in such a base class, it is better to be supported to be customized.

Reply:
Actually they can already be customized by using config_guest_params. But more specifically or in a later stage, they still can be customized using config_guest_installation_method.